File: TAF_test/JS/cypress/cypress_demo02_schulte_tables/cypress/support/page_objects/base_page.js
class BasePage {
  static BASE_URL = "https://schultetable.web.app";
  static CELL_SELECTOR = ".cell";
  static BOARD_SELECTOR = ".board-container";
  static RESTART_BUTTON = ".restart-button";
  static TOP_BAR = ".top-bar";
  static DESCRIPTION = ".desc";
  static NEXT_NUMBER = ".top-bar span";
  static TIMER = ".top-bar p:last-child";

  navigateTo(url) {
    cy.visit(url);
  }

  clickElement(selector) {
    cy.get(selector).click();
  }

  getText(selector) {
    return cy.get(selector).invoke("text");
  }

  waitForSelector(selector, timeout = 5000) {
    cy.get(selector, { timeout }).should("exist");
  }

  isElementVisible(selector) {
    return cy.get(selector).should("be.visible");
  }

  getBoardCells() {
    return cy.get(BasePage.CELL_SELECTOR);
  }

  getCellCount() {
    return this.getBoardCells().its("length");
  }

  getCellValues() {
    const values = [];
    return this.getBoardCells().each(($cell) => {
      values.push($cell.text());
    }).then(() => values);
  }

  clickCell(index) {
    this.getBoardCells().eq(index).click();
  }

  getNextNumber() {
    return cy.get(BasePage.NEXT_NUMBER).invoke("text");
  }

  clickRestart() {
    cy.get(BasePage.RESTART_BUTTON).click();
  }

  isBoardVisible() {
    return cy.get(BasePage.BOARD_SELECTOR).should("be.visible");
  }

  getDescriptionText() {
    return cy.get(BasePage.DESCRIPTION).invoke("text");
  }
}

export default BasePage;

File: TAF_test/schulte-playwright-ui-tests/page_objects/base_page.py
from logger.logger import Logger


class BasePage:
    BASE_URL = "https://schultetable.web.app"
    CELL_SELECTOR = ".cell"
    BOARD_SELECTOR = ".board-container"
    RESTART_BUTTON = ".restart-button"
    TOP_BAR = ".top-bar"
    DESCRIPTION = ".desc"
    NEXT_NUMBER = ".top-bar span"
    TIMER = ".top-bar p:last-child"

    def __init__(self, page, logger=None):
        self.page = page
        self.logger = logger or Logger()

    def navigate_to(self, url):
        self.logger.info(f"Navigating to {url}")
        self.page.goto(url)

    def click(self, selector):
        self.page.locator(selector).click()

    def get_text(self, selector):
        return self.page.locator(selector).text_content()

    def wait_for_selector(self, selector, timeout=5000):
        self.page.locator(selector).wait_for(timeout=timeout)

    def is_element_visible(self, selector):
        return self.page.locator(selector).is_visible()

    def get_board_cells(self):
        return self.page.locator(self.CELL_SELECTOR)

    def get_cell_count(self):
        return self.get_board_cells().count()

    def get_cell_values(self):
        cells = self.get_board_cells()
        count = cells.count()
        return [cells.nth(i).inner_text() for i in range(count)]

    def click_cell(self, index):
        self.get_board_cells().nth(index).click()

    def get_next_number(self):
        return self.get_text(self.NEXT_NUMBER)

    def click_restart(self):
        self.click(self.RESTART_BUTTON)

    def is_board_visible(self):
        return self.is_element_visible(self.BOARD_SELECTOR)

    def get_description_text(self):
        return self.get_text(self.DESCRIPTION)